mrmustard/utils/training.py

```diff
@@ -73,8 +73,9 @@ def minimize(
                     update_euclidean(params["euclidean"], grads["euclidean"], self.euclidean_lr)
                     self.opt_history.append(cost)
                     bar.step(math.asnumpy(cost))
-        except KeyboardInterrupt:  # graceful exit
-            return
+        except KeyboardInterrupt as e:  # graceful exit
+            print("KeyboardInterrupt: Stopping execution...")
+            raise StopExecution from e
```

Comment:
I'm not sure if this is the best way to do this. It's not that "graceful" of an exit, since another error is simply raised from the `KeyboradInterrupt` error. I get a:
```
~/miniforge3/envs/mm/lib/python3.9/site-packages/IPython/core/ultratb.py in stb2text(self, stb)
   1281     def stb2text(self, stb):
   1282         """Convert a structured traceback (a list) to a string."""
-> 1283         return self.tb_join_char.join(stb)
   1284 
   1285 

TypeError: can only join an iterable
```
with the
```
~/Repos/mrmustard/mrmustard/utils/training.py in minimize(self, cost_fn, by_optimizing, max_steps)
     77             print("KeyboardInterrupt: Stopping execution...")
---> 78             raise StopExecution from e
     79 

StopExecution: 

During handling of the above exception, another exception occurred:
```
much higher up, and the `"KeyboardInterrupt: Stopping execution..."` basically out-of-sight.

Reply:
Some other possible options:
* Simply remove the try-except block and thus raise the first `KeyboardInterrupt` error.
* Use `raise SystemExit("KeyboardInterrupt: Stopping execution...")` (or `sys.exit("KeyboardInterrupt: Stopping execution...")`), although I'm not sure about this one, since it'll basically work like the current solution.
* Have `minimize` return a status (e.g., a "finished" bool, or a status message, e.g., "success", "failed", "break") which then needs to be used in the outer for-loop:
```python
# The for loop is the important part. The exact circuit is irrelevant.
for _ in range(10):
    opt = Optimizer(symplectic_lr=0.1, euclidean_lr=0.01)
    status = opt.minimize(cost_fn_eucl, by_optimizing=[D])
    if status == "break":
        break
```
I find the latter perhaps the nicest solution, although it wouldn't solve the posted issue directly, but rather change the accepted syntax.

Reply:
Are there any advantages to having the "graceful exit"? I'm fine with an error being printed when I use KeyboardInterrupt.

Reply:
The usual reasons for having a graceful exit is to be able to keep running code after the exit has been made, e.g., the example you had where the loop kept running even though a `KeyboardInterrupt` was raised (if that would have been advantageous). I reckon an example where this could be used is when you wish to abort an optimization but still keep the the data/minimized value(s) up to that point, to be able to save or plot it. Potentially useful if it's a long optimization and you don't want to discard the results just because it hasn't finished yet.

Reply:
@thisac could this be solved by having 

```python
except KeyboardInterrupt:
    raise OptimizerInterruptedError("some message") from None
```

from what I recall, the `from None` suppresses the previous traceback.

On another note, it is always best practice not to include hardcoded `print` statements in libraries, since they are not configurable by users. I suggest either changing it to an error message, or using a `logger`.

Reply:
Thanks @josh146. I think that might be the best solution here.

Reply:
Thanks all for your comments!